[BEAM-1763] Verify PAssert execution in runners which support metrics.

aviemzur opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

aviemzur commented:
Run Spark ValidatesRunner

aviemzur commented:
Run Flink ValidatesRunner

aviemzur commented:
Run Dataflow ValidatesRunner

aviemzur commented:
Run Apex ValidatesRunner

aviemzur commented:
R: @tgroh / @kennknowles 
CC: @aljoscha @amitsela @pabloem 

@aljoscha Flink runner has several failures that are surfaced by these validations that need fixing. We can probably reuse fixes implemented in https://github.com/apache/beam/pull/2263.

aviemzur commented:
Run Spark ValidatesRunner

aviemzur commented:
Run Flink ValidatesRunner

aviemzur commented:
Run Dataflow ValidatesRunner

aviemzur commented:
Run Apex ValidatesRunner

kennknowles commented on sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java:
```diff
@@ -266,14 +273,45 @@ public void visitPrimitiveTransform(Node node) {
     try {
-      return runner.run(this);
+      PipelineResult pipelineResult = runner.run(this);
+      verifyPAssertsSucceeded(pipelineResult);
```
`Pipeline.run()` is asynchronous and the `PipelineResult` may often just be a wrapper on a job id, so you can't verify here.

aviemzur replied:
Right. Will move this to `TestPipeline`

aviemzur commented:
Retest this please

aljoscha commented:
@jkff `SplittableDoFnTest.testOutputAfterCheckpoint()` fails on the Flink Runner. If you instrument the test (with a little sysout printing in a ParDo) you see that the SDF only emits elements up to 12344. If, in `BoundedSourceWrapper` (which executes the `Create.of("foo")` on Flink), you replace the end of the `run()` method by
```
    // emit final Long.MAX_VALUE watermark, just to be sure
    // ctx.emitWatermark(new Watermark(Long.MAX_VALUE));

    while (isRunning) {
      Thread.sleep(500);
    }
```
you see all values correctly emitted but the test never stops because the source is now unbounded. (Interestingly, if you replace that one line by `ctx.emitWatermark(new Watermark(BoundedWindow.TIMESTAMP_MAX_VALUE.getMillis()));` you get values up to 34567)


The problem seems to be that the code running the SDF sees the high watermarks and then doesn't emit stuff anymore. Do you have an idea what could be going on there? If not we'll probably have to delve deeper into the code.

aljoscha commented:
Run Flink ValidatesRunner

jkff commented:
@aljoscha SDF code does not inspect watermarks, it only sets watermark holds. For further debugging I'd recommend to add some instrumentation in OutputAndTimeBoundedSplittableProcessElementInvoker, as well as in the fn under test itself.

It is expected, in this test, with default settings of that invoker (which I don't think Flink changes), that it will output up to 12344 in one invocation of ProcessElement, and output the rest in subsequent invocations. If it's getting stalled, that means the continuation is not getting scheduled - e.g. perhaps because the watermark hold didn't get set or got lost, and Flink simply dropped the continuation as "late data".

jkff commented:
See also ProcessFn.processElement which sets the watermark holds.

aviemzur commented:
Run Dataflow ValidatesRunner

aviemzur commented:
Run Dataflow ValidatesRunner

aviemzur commented:
Run Flink ValidatesRunner

aviemzur commented:
Run Apex ValidatesRunner

aviemzur commented:
Run Spark ValidatesRunner

aljoscha commented:
What's the status of this, now that the Flink test got "fixed"?

aviemzur commented:
Looks like there are failures in other tests now, perhaps these are flaky? @aljoscha can you take a look at the Flink ValidatesRunner that failed in the checks section?

aviemzur commented:
Dataflow failure is unrelated as the build has been broken there for a while.

What's left to do: 
1. Figure out why Flink tests fail (flaky?) @aljoscha 
2. Since we moved this from `Pipeline` to `TestPipeline` as discussed with @kennknowles some of the Spark unit tests don't go through this verification because they do not use `TestPipeline`, I have a solution for this locally and will discuss with @amitsela

aljoscha commented:
Run Flink ValidatesRunner

aljoscha commented:
It's hard to know why the tests fail because it just says that only 2 out of 3 expected assertions succeeded. I think we have to add what @kennknowles once suggested: to give each assertion a unique ID and to listen on all assertions succeeding instead of just counting them.

kennknowles commented:
I'm a little bit lost about the unique ID bit. I meant to have a separate downstream transform listening to each one. I wonder if your idea is even better?

aljoscha commented:
@kennknowles I'm basically talking about something like this: https://github.com/apache/flink/blob/9ed98f2e5a32fb14de03b9a8ea1dd45851cc3a7e/flink-tests/src/test/java/org/apache/flink/test/checkpointing/utils/StatefulJobSavepointFrom12MigrationITCase.java#L153-L153

Each operator does one verification and updates a unique accumulator. The `restoreAndExecute()` listens on the accumulators and waits until each one of them is reported.

In Beam, the operators that update the Accumulator (or Metric) would be `PAssert` operators that each have a unique ID for their assert. The `TestPipeline` would then listen on the IDs that it can get using pipeline traversal.

aljoscha commented:
I consider the unique-id-thing future work, by the way. I would even be in favor of merging it as is and trying to figure out what the Flink failures are as we go. The tests are not broken by these additions, we just get a better signal and that’s telling us that there seems to be some flakiness.

What do you think?

aviemzur commented:
Makes sense to do it as future work. Can you open a ticket for what you were thinking of?

aviemzur commented:
Run Spark ValidatesRunner

aviemzur commented:
R: @amitsela PTAL at the latest commit, removed Spark runner's `PipelineRule` in favor of `TestPipeline` (since that is where the `PAssert` validation happens) and added a Surefire plugin execution for streaming tests (similarly to what is done in Flink runner) . Wanted to make sure there was no reason to keep `PipelineRule`.

@kennknowles made some minor changes to `TestPipeline` PTAL.

kennknowles commented:
Last time I checked, Dataflow `ValidatesRunner` was the only one that failed - was it just `AvroCoderTest` or was it more than that?

kennknowles commented on runners/spark/src/test/java/org/apache/beam/runners/spark/ProvidedSparkContextTest.java:
```diff
@@ -95,7 +97,9 @@ private void testWithValidProvidedContext(JavaSparkContext jsc) throws Exception
-        p.run().waitUntilFinish();
+        PipelineResult result = p.run();
+
+        TestPipeline.verifyPAssertsSucceeded(p, result);
```
Any reason not to make this an instance method like `verifyPAssertsSucceeded()`?

aviemzur replied:
This specific test is a hassle because it uses `Pipeline` and not `TestPipeline`.

kennknowles commented on runners/spark/src/test/java/org/apache/beam/runners/spark/StreamingTest.java:
```diff
@@ -0,0 +1,23 @@
+/**
+ * Category tag for tests that should be run in streaming mode.
+ */
+public interface StreamingTest {}
```
Just a passing note that there's already a `UsesUnboundedPCollection` in the core SDK. It may be slightly different and no need to change now.

aviemzur replied:
That could work, however semantically these aren't tests which use unbounded `PCollection`s, rather bounded `PCollections` which we adapt to unbounded ones by setting `forceStreaming` in `TestSparkPipelineOptions` when running these tests.

aviemzur commented:
Yeah it was `AvroCoderTest`

kennknowles commented on sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java:
```diff
@@ -445,7 +445,7 @@ OutputT applyTransform(String name, InputT input,
   private CoderRegistry coderRegistry;
   private final List<String> unstableNames = new ArrayList<>();
-  private final PipelineOptions defaultOptions;
+  protected final PipelineOptions defaultOptions;
```
Not a fan of protected access to fields. In #2857 I simply store the options in the `TestPipeline` to keep them a smidge more decoupled. You can leave this; I'll rebase and adjust.

aviemzur replied:
👍

aviemzur commented:
Retest this please.

kennknowles commented on sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java:
```diff
@@ -238,7 +245,7 @@ protected void afterUserCodeFinished() {
   }
 
-  static final String PROPERTY_BEAM_TEST_PIPELINE_OPTIONS = "beamTestPipelineOptions";
+  public static final String PROPERTY_BEAM_TEST_PIPELINE_OPTIONS = "beamTestPipelineOptions";
```
It makes sense that actually both of these properties should be public, since they reach all the way to the world outside Java.

Javadoc?

aviemzur replied:
👍

kennknowles commented on sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java:
```diff
@@ -369,6 +377,15 @@ public TestPipeline enableAutoRunIfMissing(final boolean enable) {
+  /**
+   * Get this pipeline's options.
+   */
+  public PipelineOptions getOptions() {
```
In #2857 I also added this, while removing it from the superclass.

aviemzur replied:
👍

kennknowles commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SideInputHandler.java:
```diff
@@ -161,11 +162,6 @@ public void addSideInputValue(
   @Override
   public <T> T get(PCollectionView<T> sideInput, BoundedWindow window) {
 
-    if (!isReady(sideInput, window)) {
```
What is this deletion about?

aviemzur replied:
Looks like @tgroh asked @aljoscha this same question in the [original PR](https://github.com/apache/beam/pull/2263#pullrequestreview-27893521)
@aljoscha ?

kennknowles commented:
run flink validatesrunner

kennknowles commented:
run dataflow validatesrunner

aviemzur commented:
Run Spark Validates Runner

aviemzur commented:
Run Dataflow Validates Runner

aviemzur commented:
Run Flink Validates Runner

aviemzur commented:
Run Apex Validates Runner

aviemzur commented:
Run Spark ValidatesRunner

aviemzur commented:
Run Flink Validates Runner

aviemzur commented:
Run Dataflow Validates Runner

aviemzur commented:
Run Apex Validates Runner

aviemzur commented:
Run Flink ValidatesRunner

aviemzur commented:
Run Dataflow ValidatesRunner

aviemzur commented:
Run Apex ValidatesRunner

amitsela commented on runners/spark/src/test/java/org/apache/beam/runners/spark/CacheTest.java:
```diff
@@ -36,12 +36,12 @@
-    Pipeline pipeline = pipelineRule.createPipeline();
+    SparkPipelineOptions options =
+        PipelineOptionsFactory.create().as(TestSparkPipelineOptions.class);
+    options.setRunner(SparkRunner.class);
```
`TestSparkRunner` ?

aviemzur replied:
👍

amitsela commented on runners/spark/src/test/java/org/apache/beam/runners/spark/ForceStreamingTest.java:
```diff
@@ -44,19 +45,24 @@
+    pipeline.apply(GenerateSequence.from(0));
 
     // apply the BoundedReadFromUnboundedSource.
     BoundedReadFromUnboundedSource<?> boundedRead =
```
Is this still necessary if you're using `GenerateSequence` ?

aviemzur replied:
👍 will remove

amitsela commented on runners/spark/src/test/java/org/apache/beam/runners/spark/translation/streaming/ResumeFromCheckpointStreamingTest.java:
```diff
@@ -96,24 +99,34 @@
  * expected to recover as well.
  */
-public class ResumeFromCheckpointStreamingTest {
+public class ResumeFromCheckpointStreamingTest implements Serializable {
```
Scala Serializable ? probably want Java's..

aviemzur replied:
👍
